addons/mass_mailing/models/mailing_list.py

```diff
@@ -195,68 +195,110 @@ def action_view_contacts_bouncing(self):
+        self.env['mailing.contact'].flush_model()
+        contact_condition = self._mailing_list_get_contact_condition()
+        select_values = [f"contact.{value}" for value in self._mailing_list_get_select_values()]
+        select_values_string = ', '.join(select_values)
+        select = ', '.join(f"COALESCE({field}, '')" for field in select_values)
+        join_select = ', '.join(f"COALESCE(sub_{field}, '')" for field in select_values) #Need to add coalesce as in postgresql (NULL, 1) = 1
+        insert_query = f"""
             INSERT INTO mailing_contact_list_rel (contact_id, list_id)
-            SELECT st.contact_id AS contact_id, %s AS list_id
-            FROM
-                (
-                SELECT
-                    contact.id AS contact_id,
-                    contact.email AS email,
-                    list.id AS list_id,
-                    row_number() OVER (PARTITION BY email ORDER BY email) AS rn
-                FROM
-                    mailing_contact contact,
-                    mailing_contact_list_rel contact_list_rel,
-                    mailing_list list
-                WHERE contact.id=contact_list_rel.contact_id
-                AND COALESCE(contact_list_rel.opt_out,FALSE) = FALSE
-                AND contact.email_normalized NOT IN (select email from mail_blacklist where active = TRUE)
+            SELECT DISTINCT ON ({select_values_string})
+                contact.id AS contact_id,
+                %(id)s AS list_id
+              FROM mailing_contact contact,
+                  mailing_contact_list_rel contact_list_rel,
+                  mailing_list list
+             WHERE contact.id=contact_list_rel.contact_id
+                AND {contact_condition}
                 AND list.id=contact_list_rel.list_id
-                AND list.id IN %s
-                AND NOT EXISTS
-                    (
-                    SELECT 1
-                    FROM
-                        mailing_contact contact2,
-                        mailing_contact_list_rel contact_list_rel2
-                    WHERE contact2.email = contact.email
-                    AND contact_list_rel2.contact_id = contact2.id
-                    AND contact_list_rel2.list_id = %s
+                AND list.id IN %(ids)s
+                AND ({select}) NOT IN (
+                      SELECT {join_select}
+                        FROM mailing_contact_list_rel rel
+                        JOIN mailing_contact sub_contact ON rel.contact_id=sub_contact.id
+                       WHERE rel.list_id = %(id)s
+                );
+        """
+        update_query = f"""
+            UPDATE mailing_contact_list_rel
+               SET opt_out = TRUE
+             WHERE list_id = %(id)s
+               AND contact_id in (
+                    SELECT id
+                      FROM mailing_contact contact
+                     WHERE ({select}) IN (
+                            SELECT {select.replace("contact.", "")}
```

Comment:
how this work ?
You will have 
```
WHERE COALESCE('email', ''), COALESCE('phone', '') IN (
    SELECT COALESCE('email', ''), COALESCE('phone', '')
      FROM ...
```

it take the record if one of the 2 field match, only if both match, etc ? (my intuition tell me that if will take it only if both match, not 100% sure)

could be interresting to add a comment in the query to explain it (`-- comment`)

Reply:
@std-odoo  Actually I guess there's an issue with that in this example:

```
phone | email
123     | t@t.lan
123     | t2@t.lan
```
What should happen here? Phone seems to indicate it's the same contact, but we can't just delete an email.

So maybe that was actually intended (i.e contacts are only the same if all data matches)?

Reply:
> What should happen here? Phone seems to indicate it's the same contact, but we can't just delete an email.

Indeed, maybe it's better to consider the contact is the same if both email and phone are the same

> So maybe that was actually intended (i.e contacts are only the same if all data matches)?

Yes maybe, could be worth to add a comment to make it explicit